### PR TITLE
add optimze of dsv3

### PR DESCRIPTION
**Pull request by momo609**

### CODE REVIEW

File: vllm_ascend/models/deepseek_v2.py [hunk 1]
```diff
@@ -224,8 +224,7 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
             enable_force_load_balance = False
         num_tokens, hidden_dim = hidden_states.shape
 
-        if self.n_shared_experts is not None:
-            shared_output = self.shared_experts(hidden_states)
+        old_hidden_states = hidden_states.detach()
```
Comment (MengqingCao) [inline vllm_ascend/models/deepseek_v2.py]: Why detach hidden states here?

File: vllm_ascend/ops/fused_moe.py [hunk 2]
```diff
@@ -371,7 +371,7 @@ def fused_experts(
                             num_experts)).to(topk_ids.dtype)
 
         # Sort by local expert IDs
-        sort_indices = torch.argsort(filtered_experts)
+        sort_indices = torch.argsort(filtered_experts.view(torch.float32))
```
Comment (MengqingCao) [inline vllm_ascend/ops/fused_moe.py:365]: why we change the dtype of `filterd_experts` to float32? And `view` will change the metadata of `filterd_experts`, instead of create a new tensor. Is this expected?

File: vllm_ascend/ops/utils.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from typing import Callable, Optional
+from vllm.model_executor.layers.utils import dispatch_unquantized_gemm, rocm_unquantized_gemm
+from vllm.platforms import current_platform
+
+
+def npu_matmul_add(x: torch.Tensor,
+                weight: torch.Tensor,
+                bias: Optional[torch.Tensor] = None):
+    import torch_npu
+    if torch_npu.get_npu_format(weight) != 29:
```
Comment (MengqingCao) [inline vllm_ascend/ops/utils.py]: What format does 29 refer to? Let's add a comment on it.

File: vllm_ascend/ops/utils.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        out = torch.empty((n, m), dtype=x.dtype, device=x.device)
+        torch_npu._npu_matmul_add_fp32(x, weight, out)
+        return out
+
+def dispatch_unquantized_gemm() -> Callable[..., torch.Tensor]:
+    if current_platform.is_rocm():
+        return rocm_unquantized_gemm
+    return npu_matmul_add
+
+unquantized_gemm = dispatch_unquantized_gemm
```
Comment (MengqingCao) [inline vllm_ascend/ops/utils.py]: I guess you want to patch `vllm.model_executor.layers.utils.dispatch_unquantized_gemm` into a custom one?

If so, let's do this in `vllm_ascend/patch` and discribe the details in `vllm_ascend/patch/__init
__.py`

File: vllm_ascend/sample/sampler.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    The logits tensor may be updated in-place.
+    """
+    logits = _apply_top_k_top_p(logits, k, p)
+    probs = logits.softmax(dim=-1, dtype=torch.float32)
+    return random_sample(probs, generators)
+
+s1.apply_min_p = apply_min_p
+if envs.VLLM_ENABLE_TOPK_OPTIMZE:
+    TopKTopPSampler.forward_native = topk_topp_forward_native
```
Comment (MengqingCao) [inline vllm_ascend/sample/sampler.py]: ditto. let's do this in `vllm_ascend/patch` and discribe the details in `vllm_ascend/patch/__init __.py`

File: vllm_ascend/sample/sampler.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+) -> torch.Tensor:
+    """
+    Filters logits using adaptive probability thresholding.
+    """
+    if min_p == 0:
+        return logits
+    # Convert logits to probability distribution
+    probability_values = torch.nn.functional.softmax(logits, dim=-1)
+    # Calculate maximum probabilities per sequence
+    max_probabilities = torch.amax(probability_values,
```
Comment (MengqingCao) [inline vllm_ascend/sample/sampler.py]: Does `torch.nn.functional.softmax` and `torch.amax` bring any performance gain compared with `torch.softmax` and `torch.tensor.max`?

File: vllm_ascend/sample/sampler.py (hunk 5 above)
Comment (MengqingCao) [reply vllm_ascend/sample/sampler.py]: Please add percision ut to check if the `_apply_top_k_top_p` and `apply_min_p` calculate correctly

**Review by MengqingCao — COMMENTED** [review]
1. Please add pr description
2. Run `bash format.sh` locally to fix lint failures

File: vllm_ascend/ops/fused_moe.py (hunk 2 above)
Comment (momo609) [reply vllm_ascend/ops/fused_moe.py:365]: sort can use aicore under float32, with better performance

File: vllm_ascend/models/deepseek_v2.py (hunk 1 above)
Comment (momo609) [reply vllm_ascend/models/deepseek_v2.py]: Prevent hidden_status from being overwritten by subsequent modifications。

**Comment by momo609** [comment]
@wangxiyuan

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
+#
```
Comment (wangxiyuan) [inline vllm_ascend/patch/worker/patch_common/patch_sampler.py:1]: this is a patch, so please update the doc in `__init__` as well  https://github.com/vllm-project/vllm-ascend/blob/main/vllm_ascend/patch/__init__.py

File: tests/sample/test_sampler.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (wangxiyuan) [inline tests/sample/test_sampler.py]: do not create a new sample folder, this kind of test won't be ran by defaut. Please move to singcard folder

File: vllm_ascend/models/deepseek_v2.py [hunk 9]
```diff
@@ -225,8 +225,7 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
```
Comment (ganyi1996ppo) [inline vllm_ascend/models/deepseek_v2.py]: Can you upload the performance profiling on this part? Distach this variable from the pytorch graph seems can't actually triggers the parallel execution according to my knowledge.

File: tests/e2e/doctests/001-quickstart-test.sh [hunk 10]
```diff
@@ -47,9 +47,17 @@ function quickstart_online_test() {
...
+  python3 "${SCRIPT_DIR}/../../examples/offline_inference_npu.py"
+}
+
 _info "====> Start simple_test"
 simple_test
 _info "====> Start quickstart_offline_test"
 quickstart_offline_test
 _info "====> Start quickstart_online_test"
 quickstart_online_test
+_info "====> Start quickstart_offline_test_topk"
```
Comment (Yikun) [inline tests/e2e/doctests/001-quickstart-test.sh]: We should add offline test in py by setting os.eviron rather than here:

https://github.com/vllm-project/vllm-ascend/blob/main/tests%2Fsinglecard%2Ftest_offline_inference.py#L44-L44

File: vllm_ascend/models/deepseek_v2.py (hunk 9 above)
Comment (momo609) [reply vllm_ascend/models/deepseek_v2.py]: The original moe is originally on two streams, so we only need to change the calculation position of the shared expert.

File: tests/singlecard/test_offline_inference.py [hunk 11]
```diff
@@ -81,3 +82,27 @@ def test_multimodal(model, prompt_template, vllm_runner):
         vllm_model.generate_greedy(prompts=prompts,
                                    images=images,
                                    max_tokens=64)
+
+
+@pytest.mark.parametrize("model", MODELS)
+@pytest.mark.parametrize("dtype", ["half", "float16"])
+@pytest.mark.parametrize("max_tokens", [5])
```
Comment (wangxiyuan) [inline tests/singlecard/test_offline_inference.py]: don't need the full test here, just one is enough.

**Review by wangxiyuan — COMMENTED** [review]
please add a ds test for multicard as well. example: https://github.com/vllm-project/vllm-ascend/blob/main/tests/multicard/test_offline_inference_distributed.py#L49

File: tests/e2e/doctests/001-quickstart-test.sh (hunk 10 above)
Comment (wangxiyuan) [reply tests/e2e/doctests/001-quickstart-test.sh]: ++

File: tests/multicard/test_offline_inference_distributed.py [hunk 12]
```diff
@@ -61,3 +62,23 @@ def test_models_distributed_DeepSeek():
             distributed_executor_backend="mp",
     ) as vllm_model:
         vllm_model.generate_greedy(example_prompts, max_tokens)
+
+
+def test_models_distributed_topk(model) -> None:
+    os.environ["VLLM_ENABLE_TOPK_OPTIMZE"] = "1"
```
Comment (Yikun) [inline tests/multicard/test_offline_inference_distributed.py]:
```suggestion
@patch.dict(os.environ, {"VLLM_ENABLE_TOPK_OPTIMZE": "1"})
def test_models_distributed_topk(model) -> None:
```

we should unset VLLM_ENABLE_TOPK_OPTIMZE after test executed

[1] https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.dict

File: tests/multicard/test_offline_inference_distributed.py [hunk 13]
```diff
@@ -25,6 +25,7 @@
 import vllm  # noqa: F401
```
Comment (Yikun) [inline tests/multicard/test_offline_inference_distributed.py:26]: import os
from unittest.mock import patch

import vllm

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    The logits tensor may be updated in-place.
+    """
+    logits = _apply_top_k_top_p(logits, k, p)
+    probs = logits.softmax(dim=-1, dtype=torch.float32)
+    return random_sample(probs, generators)
+
+
+Sampler.apply_min_p = apply_min_p
+if envs.VLLM_ENABLE_TOPK_OPTIMZE:
+    TopKTopPSampler.forward_native = topk_topp_forward_native
```
Comment (Yikun) [inline vllm_ascend/patch/worker/patch_common/patch_sampler.py]: newline

**Comment by Yikun** [comment]
Please also update PR title and message more accuracy and specific, `add optimze of dsv3` is more general

File: vllm_ascend/ops/fused_moe.py (hunk 2 above)
Comment (momo609) [reply vllm_ascend/ops/fused_moe.py:365]: sort can use aicore under float32, with better performance

File: vllm_ascend/ops/utils.py (hunk 3 above)
Comment (momo609) [reply vllm_ascend/ops/utils.py]: fixed

File: vllm_ascend/ops/utils.py (hunk 4 above)
Comment (momo609) [reply vllm_ascend/ops/utils.py]: fixed

File: vllm_ascend/sample/sampler.py (hunk 5 above)
Comment (momo609) [reply vllm_ascend/sample/sampler.py]: fixed

File: vllm_ascend/sample/sampler.py (hunk 6 above)
Comment (momo609) [reply vllm_ascend/sample/sampler.py]: This function changes indexput to masked_fill to achieve performance optimization. The operation in the comment is not modified.

File: tests/multicard/test_offline_inference_distributed.py (hunk 13 above)
Comment (momo609) [reply tests/multicard/test_offline_inference_distributed.py:26]: fixed

File: vllm_ascend/patch/worker/patch_common/patch_sampler.py (hunk 7 above)
Comment (momo609) [reply vllm_ascend/patch/worker/patch_common/patch_sampler.py:1]: fixed

File: tests/singlecard/test_offline_inference.py [hunk 15]
```diff
@@ -81,3 +83,24 @@ def test_multimodal(model, prompt_template, vllm_runner):
         vllm_model.generate_greedy(prompts=prompts,
                                    images=images,
                                    max_tokens=64)
+
+
+@patch.dict(os.environ, {"VLLM_ENABLE_TOPK_OPTIMZE": "1"})
```
Comment (Yikun) [inline tests/singlecard/test_offline_inference.py]:
```suggestion
@patch.dict(os.environ, {"VLLM_ASCEND_ENABLE_TOPK_OPTIMZE": "1"})
```

File: vllm_ascend/envs.py [hunk 16]
```diff
@@ -36,6 +36,8 @@
     lambda: bool(int(os.getenv("COMPILE_CUSTOM_KERNELS", "1"))),
     "VLLM_ENABLE_MC2":
     lambda: bool(int(os.getenv("VLLM_ENABLE_MC2", '0'))),
+    "VLLM_ENABLE_TOPK_OPTIMZE":
+    lambda: bool(int(os.getenv("VLLM_ENABLE_TOPK_OPTIMZE", '0'))),
```
Comment (Yikun) [inline vllm_ascend/envs.py]:
```suggestion
    "VLLM_ASCEND_ENABLE_TOPK_OPTIMZE":
    lambda: bool(int(os.getenv("VLLM_ASCEND_ENABLE_TOPK_OPTIMZE", '0'))),
```
